rfcs/text/formik.md

```diff
@@ -0,0 +1,444 @@
+
+* **Less code to maintain**  –  Much time and effort can be saved by using a third-party library which is kept
+  well-maintained, thus freeing up one's development resources to focus on higher-priority project-unique functionality.
+* **Robustness**  –  A popular well-maintained library is likely to be thoroughly tested and debugged, especially for
+  edge cases which have been encountered and addressed by the library's community.
+* **Familiarity**  –  Users are potentially already familiar with popular third-party libraries.
+
+Considering the reasons why "reinventing the wheel" (ie: building a custom solution) can sometimes be beneficial in
+certain situations:
+
+* It can be a very educational experience for developers, but that isn't a relevant factor for this project.
+* A third-party library may be an unnecessarily heavy solution, if only a small percentage of its functionality will be
+  used by the application, and a custom-built solution may be much more lightweight.  But in this case, Formik is
+  [a relatively small library](https://jaredpalmer.com/formik/docs/overview)
+  ([14.7kB when minified and gzipped](https://bundlephobia.com/result?p=formik@2.0.4)), and we would be using a
+  significant percentage of its functionality.
+
+
+# Detailed design
+
+## Overview of Carbon's form and validation functionality
+
+Carbon provides the following components:
+
+* [**`<Form>`**](src/__experimental__/components/form/form.component.js)  –  This augments an HTML
+  [`<form>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form) element with: a footer (containing "Save"
+  and "Cancel" buttons), form submission functionality, [CRSF](https://en.wikipedia.org/wiki/CRSF) functionality, a
+  [`beforeunload`](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event) "unsaved" warning, etc.
+* [**`<FormField>`**](src/__experimental__/components/form-field/form-field.component.js)  –  This enables a
+  [`<Label>`](src/__experimental__/components/label/label.component.js) and
+  [`<FieldHelp>`](src/__experimental__/components/field-help/field-help.component.js) to be added to an individual
+  form field.
+* [**`withValidation`**](src/components/validations/with-validation.hoc.js)  –  This HOC wraps an individual form field,
+  enabling validation functions (for errors, warnings and info) to be executed when the component blurs or receives
+  updated props.
+
+
+## Overview of Formik
+
+> Formik keeps track of your form's state, and then exposes it (plus a few reusable methods and event handlers) to your
+> form via props. [🔗](https://jaredpalmer.com/formik/docs/overview#the-gist)
+
+Formik provides a small set of React components and hooks, of which the key ones are:
+
+* [**`<Formik>`**](https://jaredpalmer.com/formik/docs/api/formik)  –  The central component of Formik, which wraps an
+  HTML form, and accepts props such as
+  [`initialValues`](https://jaredpalmer.com/formik/docs/api/formik#initialvalues-values),
```

Comment:
We will probably want to document somewhere that Formik will only work for controlled inputs, highly unlikely that someone would try to use it with uncontrolled but better if we make it explicit IMO

Reply:
Good point :thumbsup: I suppose technically it's possible to use Formik with an uncontrolled input, by wiring up [`handleChange`](https://jaredpalmer.com/formik/docs/api/formik#handlechange-e-reactchangeevent-any-void) (for Formik's validation functionality) but not wiring up `value` to Formik. But that would cause complications, and it's not really how Formik is intended to be used.

Reply:
Yeah good point, think we are best not encouraging it's use uncontrolled but we can't really force the hand of developers using Carbon

Reply:
Added documentation about uncontrolled components in 3516b7f :thumbsup: